utils/log_parser.py

```diff
@@ -0,0 +1,311 @@
+                       "PaleTurquoise2", "PeachPuff1",   "PowderBlue", "Salmon",
+                       "Thistle1",       "Thistle3",     "Wheat1",     "Yellow2"]
+
+    def get_unique_vertex_name(self, node):
+        if node not in self.uniqueVertexMap:
+            self.uniqueVertexMap[node] = self.uniqueVertexNo
+            self.uniqueVertexNo += 1
+        return "v"+str(self.uniqueVertexMap[node])
+
+    def dump_label(self, node):
+        labelStr = "{"
+        labelStr += " {<Inputs>Inputs}|"
+        labelStr += "{" + node.get_kind_name() + "\lname: " + \
+            node.get_name()+"}|"
+        labelStr += " {<Outputs>Outputs}"
+        labelStr += "}"
+        return labelStr
+
+    def get_color(self, node):
+        idx = hash(node.get_kind_name()) % len(self.colors)
+        return self.colors[idx]
+
+    def dump_node(self, node):
+        if not node:
+            return
+        nodeStr = self.get_unique_vertex_name(node) + "[\n"
+        nodeStr += "\tlabel = \""
+        nodeStr += self.dump_label(node)
+        nodeStr += "\"\n"
+        nodeStr += "\tshape = \"record\"\n"
+        nodeStr += "\tstyle=\"filled,rounded\"\n"
+        nodeStr += "\tfillcolor="
+        nodeStr += self.get_color(node)
+        nodeStr += "\npenwidth = 2];\n"
+        self.vertices_.append(nodeStr)
+
+    def visitNodes(self):
+        for k, n in NODES_MAP:
+            self.dump_node(NODES_MAP[(k, n)])
+
+    def visitEdges(self):
+        for k, n in NODES_MAP:
+            node = NODES_MAP[(k, n)]
+            for i, rn in node.get_inputs():
+                if (i.get_kind_name(), i.get_name()) not in NODES_MAP:
+                    print(i.get_kind_name(), i.get_name())
+                edgeStr = self.get_unique_vertex_name(i)+":Outputs -> "
+                edgeStr += self.get_unique_vertex_name(node)+":Inputs"
+                self.edges_.append(edgeStr)
+
+    def dump_graph(self, dagName):
+        self.visitNodes()
+        self.visitEdges()
+        f = open(dagName+"_dotty.dot", "w")
+        f.write("digraph DAG {\n\trankdir=TB;\n")
+        for v in self.vertices_:
+            f.write(v+"\n")
+        for e in self.edges_:
+            f.write(e+";\n")
+        f.write("}")
+        f.close()
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Parse compilation log")
+    parser.add_argument("-f", "--log-file")
+    parser.add_argument("--dump-phases", nargs="+", choices=CRITICAL_PHASES)
+    options = parser.parse_args()
+
+    if options.dump_phases:
+        dumpPhases = options.dump_phases
+    else:
+        dumpPhases = CRITICAL_PHASES
+
+    return options.log_file, dumpPhases
+
+
+def dump_dag(dagName):
+    dotty = DottyPrinter()
+    dotty.dump_graph(dagName)
+
+
+def process(logLines, dumpPhases):
+    phaseNo = 0
+    for ln in logLines:
+        # Process CREATE statement
+        if re.search("CREATE", ln):
+            p = re.compile(
```

Comment:
You probably want to hoist the compilation of a regex out of the loop. It will make things significantly faster.